### PR TITLE
fix(config_list): Add JSON output

### PR DESCRIPTION
**Pull request by salome-voltz**
## What has been done

This MR contains two commits: 
-  [feat(config_list): Add JSON output] : Adds JSON output to command `ggshield config list`
- [fix(api-status): Update JSON output] : Updates JSON output for command `ggshield api-status` to include the instance which is checked. (This info is present in the text output). 

Also, [some stages name for pre-commit have been changed](https://pre-commit.com/#confining-hooks-to-run-at-certain-stages).
 
## PR check list

- [x] As much as possible, the changes include tests (unit and/or functional)
- [x] If the changes affect the end user (new feature, behavior change, bug fix) then the PR has a changelog entry (see doc/dev/getting-started.md). If the changes do not affect the end user, then the `skip-changelog` label has been added to the PR.

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (agateau-gg) [inline .pre-commit-config.yaml:1]: The changes in this file have nothing to do with adding JSON output 😅. They should at least be in their own commit.

File: ggshield/cmd/status.py [hunk 1]
```diff
@@ -29,8 +30,12 @@ def status_cmd(ctx: click.Context, **kwargs: Any) -> int:
     if not isinstance(response, HealthCheckResponse):
         raise UnexpectedError("Unexpected health check response")
 
+    # Add the API URL to the JSON output
+    json_output = response.to_dict()
+    json_output["instance"] = client.base_uri
+
     click.echo(
-        response.to_json()
+        json.dumps(json_output, indent=2)
```
Comment (gg-jonathangriffe) [inline ggshield/cmd/status.py:38]: It would maybe be better to change `HealthCheckResponse` and `health_check` to keep `response.to_json()` here.

File: tests/unit/cmd/test_status.py [hunk 2]
```diff
@@ -36,6 +36,7 @@ def test_api_status(cli_fs_runner, api_status_json_schema):
             All(
                 {
                     "detail": "Valid API key.",
+                    "instance": Match(r"https://[^\s]+\.com"),
```
Comment (gg-jonathangriffe) [inline tests/unit/cmd/test_status.py:39]: I'm not sure we want to match the `.com` here as it will fail when running the tests with a non default instance.
Maybe it's fine since other tests would also fail ? WDYT ?

File: ggshield/cmd/config/config_list.py [hunk 3]
```diff
@@ -1,35 +1,37 @@
...
     Print the list of configuration keys and values.
     """
-    config = ContextObj.get(ctx).config
+    ctx_obj = ContextObj.get(ctx)
+    config = ctx_obj.config
     default_token_lifetime = config.auth_config.default_token_lifetime
 
-    message_lines = []
+    # Initialize the structure for JSON output
+    config_data = {"instances": [], "global_values": {}}
```
Comment (gg-jonathangriffe) [inline ggshield/cmd/config/config_list.py:25]: It would maybe be clearer if we use a schema here, wdyt ?

File: ggshield/cmd/status.py (hunk 1 above)
Comment (salome-voltz) [reply ggshield/cmd/status.py:38]: `HealthCheckResponse` is defined in pygitguardian. We can change it but it requires more work. 
I agree that it would have been more understandable, but I'm not entirely sure of where `HealthCheckResponse` is used. We might need to check the imports before modifying it :)

File: .pre-commit-config.yaml
Comment (salome-voltz) [reply .pre-commit-config.yaml:1]: :innocent: fixed!

File: ggshield/cmd/config/config_list.py [hunk 4]
```diff
@@ -1,35 +1,37 @@
...
 
-    def add_entries(*entries: Tuple[str, Any]):
+    def add_global_entries(*entries: Tuple[str, Any]):
         for key, value in entries:
-            message_lines.append(f"{key}: {value}")
+            config_data["global_values"][key] = value
```
Comment (agateau-gg) [inline ggshield/cmd/config/config_list.py:29]: This function is only used once: it can be inlined

File: tests/unit/cmd/test_config.py [hunk 5]
```diff
@@ -45,7 +50,7 @@
 
 
 class TestConfigList:
-    def test_valid_list(self, cli_fs_runner):
+    def test_valid_list(self, cli_fs_runner, config_list_json_schema):
```
Comment (agateau-gg) [inline tests/unit/cmd/test_config.py:53]: I would prefer if there was two separate tests: one for text and one for the JSON output. This way we have better diagnostics if one of them fails (and each test is shorter).

You can put the setup in a class-level fixture to avoid code duplication.

File: ggshield/cmd/status.py [hunk 6]
```diff
@@ -29,8 +30,12 @@ def status_cmd(ctx: click.Context, **kwargs: Any) -> int:
     if not isinstance(response, HealthCheckResponse):
         raise UnexpectedError("Unexpected health check response")
 
+    # Add the API URL to the JSON output
+    json_output = response.to_dict()
+    json_output["instance"] = client.base_uri
```
Comment (agateau-gg) [inline ggshield/cmd/status.py:35]: This code is only for the JSON output, I think it's better to split the `click.echo()` call instead. Something like this:

```python
if ctx_obj.use_json:
    json_output = response.to_dict()
    json_output["instance"] = client.base_uri
    click.echo(json.dumps(json_output))
else:
    # text output
    ...
```

File: ggshield/cmd/config/config_list.py [hunk 7]
```diff
@@ -1,35 +1,37 @@
...
-from ggshield.cmd.utils.common_options import add_common_options
+from ggshield.cmd.utils.common_options import add_common_options, json_option
 from ggshield.cmd.utils.context_obj import ContextObj
 
 from .constants import DATETIME_FORMAT, FIELDS
 
 
 @click.command()
 @click.pass_context
+@json_option
```
Comment (agateau-gg) [inline ggshield/cmd/config/config_list.py:14]: Can you also add `@text_json_format_option`? It adds the `--format` option which is now provided by all commands with a `--json` option.

File: ggshield/cmd/config/config_list.py [hunk 8]
```diff
@@ -54,16 +56,34 @@ def add_entries(*entries: Tuple[str, Any]):
...
+            "token": token,
+            "token_name": token_name,
+            "expiry": expiry,
+        }
+        config_data["instances"].append(
+            {"instance_name": instance_name, **instance_info}
         )
+
+    if ctx_obj.use_json:
+        click.echo(json.dumps(config_data, indent=2))
```
Comment (agateau-gg) [inline ggshield/cmd/config/config_list.py:72]: No need for `indent=2`: this is for machines, and non machines can pipe it to `jq` :)

**Review by agateau-gg — CHANGES_REQUESTED** [review]

**Review by agateau-gg — APPROVED** [review]
Looks good!
